More FV postprocessors changes

GiudGiud opened the pull request:
Fixes Interface(Integral/Average)VariableValuePostprocessor and enables the test for it. 
It's regolded because the previous test was not checking for the case of a using a single variable and no neighbor variable. 

Trying out a change that will simplify a lot of code in INSFV advection and even more in PINSFV. 

Following #16767 failures, we moved away from average interpolation for a couple tests. I think with the bug fix in #16917, we can revert back to average interpolation

By specifying the advected quantity interpolation method on internal flow postprocessors, we are now able to show perfect conservation on more cases among the diverging meshes.
This PR also adds a test capturing the bug fix in #16917 specifically

GiudGiud committed:
Rework FV interface variable value test to avoid stateful materials, refs #16149, refs #16767

GiudGiud committed:
Add assert to check for potential simplification of INSFVMomentumAdvection FVKernel and MooseVariableFV

GiudGiud committed:
Revert to testing average interpolation with asm pc, refs #16918

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/678682/) on 97aac63 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16936/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16936/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f7f66d77cd8ffeb56b76b57249cbef6402a4dbb7`

GiudGiud committed:
Add a test for the bug fix in #16918, finally show conservation of energy on internal boundaries
Add semicolons because C++

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/678707/) on ab8db95 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16936/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16936/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f7f66d77cd8ffeb56b76b57249cbef6402a4dbb7`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/678744/) on 6107da4 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16936/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16936/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f7f66d77cd8ffeb56b76b57249cbef6402a4dbb7`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/678809/) on 63f279e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16936/site)

This comment will be updated on new commits.

GiudGiud committed:
MooseVariableFV and FaceInfo are not synced with regards to elem. Use asm for no slip test too

lindsayad commented on modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C:
```diff
@@ -445,6 +445,7 @@ INSFVMomentumAdvection::interpolate(Moose::FV::InterpMethod m,
                   ? elem_one == &_face_info->elem() && elem_two == _face_info->neighborPtr()
                   : elem_one == _face_info->neighborPtr() && elem_two == &_face_info->elem(),
               "The determineElemOneAndTwo utility determined the wrong value for elem_is_elem_one");
+  mooseAssert(elem_is_elem_one, "FVFluxKernel elem should coincide with the faceInfo elem.");
```
Since this appears to be true, let's move it ahead of the assert on line 444, and then remove the else branch of the ternary

GiudGiud replied:
We could just delete all the ternaries and set elem_one to face_info->elem
I guess there's no rush to do that. Only if needed for optimization

lindsayad replied:
No you are right. Now that it's clearly demonstrated that `elem_one` is always `_face_info->elem()`, we should just remove the call to `Moose::FV::determineElemOneAndTwo` and then remove the associated asserts.

GiudGiud replied:
Ok I ll do that and add a block restricted INSFV test too

lindsayad commented on modules/navier_stokes/test/tests/postprocessors/tests:
```diff
@@ -26,6 +26,15 @@
     cli_args = "Mesh/inactive='diverging_mesh' Outputs/file_base='straight_INSFV' -pc_type lu -pc_factor_shift_type NONZERO"
     ad_indexing_type = 'global'
   []
+  [fv_straight_different_interpolation]
+    type = CSVDiff
+    input = conservation_INSFV.i
+    csvdiff =  straight_INSFV_different_interpolation.csv
+    abs_zero = 1e-8
+    requirement = 'The system shall be able to compute flow rates and, when using a different interpolation method between kernels and postprocessors, show approximate mass, momentum and energy conservation at internal and external boundaries of a frictionless heated straight channel with a finite volume incompressible Navier Stokes model.'
+    cli_args = "Postprocessors/mid1_momentum_x/advected_interp_method='upwind' Postprocessors/mid2_momentum_x/advected_interp_method='upwind' Postprocessors/mid1_advected_energy/advected_interp_method='upwind' Postprocessors/mid2_advected_energy/advected_interp_method='upwind' Mesh/inactive='diverging_mesh' Outputs/file_base='straight_INSFV_different_interpolation' -pc_type lu -pc_factor_shift_ty ...
+    ad_indexing_type = 'global'
+  []
```
Is this something we actually care to test? E.g. is this something we would actually want our users to do?

GiudGiud replied:
Well we dont want users to do that, but we also dont want that to fail hard like it was when the bug was there.

lindsayad replied:
Well ideally I think I would like it to "fail hard" when the postprocessor interpolation method is different than the kernels :smile: E.g. it would be great if it actually `mooseErrored`

GiudGiud replied:
oh yeah mooseErrored would have been fine!
I can delete the test if you dont like the requirement. 
This test covers the same code as other tests, it just covers one more potential bug.

lindsayad replied:
Ok I'm not actually sure what bug you're referring to...

GiudGiud replied:
In the fixup to the FV postproc PR, when face UO were not computing the material properties

lindsayad replied:
You mean the framework code I added? Material properties were being computed. The problem was that the variable values used in the neighbor material property computation were garbage. I am honestly unsure as to why only one test from your postprocessor test quite showed issues. I would have expected all tests with the internal volumetric flow rate postprocessor to be affected by the previous neighbor variable value garbage.

GiudGiud replied:
I thought the case you showed me with serial vs parallel was for material properties though. You were printing _advected_quantity right? (the weird 347 in serial)

My guess as to why only one case was wrong was that when the same scheme was used, the kernels, executed before the face UO, were causing the values to be computed. 

But when a different interpolation method was used, then values that were not initialized got queried by the Face UO, hence the garbage.

But it really is just a guess. Ok looks like we're looking at deletion for this test then.

lindsayad replied:
> I thought the case you showed me with serial vs parallel was for material properties though. You were printing _advected_quantity right? (the weird 347 in serial)

I was printing the temperature variable value from the material.

> My guess as to why only one case was wrong was that when the same scheme was used, the kernels, executed before the face UO, were causing the values to be computed.

This is a reasonable guess in theory. However, we compute user objects in an entirely different element loop than kernels. so what I would expect is that a user object without properly initialized variable data would, when computing on the first element/face, be seeing variable data from the *last* computed element/face from the kernel loop, or whatever element loop went immediately before the user object computation. And I would've expected this kind of bad behavior everywhere we were using that internal volumetric flow object. But seemingly not!

> But it really is just a guess. Ok looks like we're looking at deletion for this test then.

Yea I'm hoping that we're getting coverage on proper initialization of variable data from our different mixes of serial, parallel, and threaded testing with the "proper" way of setting `advected_interp_method` for both kernels and postprocessors. The fact that we had the one failure that we fixed with the proper variable initialization gives me some confidence that we do indeed have coverage without the mixed `advected_interp_method` test.

lindsayad reviewed (changes requested):
Overall looks good

GiudGiud committed:
Allow more than one boundary id per face info, refs #16972

GiudGiud committed:
Remove elem_one language in advection kernel, add block restriction test for INSFV

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/681483/) on 9749945 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16936/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16936/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f7f66d77cd8ffeb56b76b57249cbef6402a4dbb7`

GiudGiud committed:
Style changes
Remove test with a different inteprolation method in the kernel and the postprocessor

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/681508/) on 05d1086 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16936/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16936/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f7f66d77cd8ffeb56b76b57249cbef6402a4dbb7`

GiudGiud commented:
Added a partial fix to #16972. Basically if we had more than one sideset on a boundary and the first one wasnt the one holding the INSFVBC, then the kernel would not find it. Now we just loop over the sidesets till we find the one we need. 

Note that both before and after this change, a user can wrongly specify two INSFVBCs on two sidesets covering the same boundary, a full fix to #16972 will check on that particular case.

GiudGiud committed:
Restrict test to replicated meshes since it uses a PatchSideSetGenerator

lindsayad commented on modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C:
```diff
@@ -449,13 +448,6 @@ INSFVMomentumAdvection::interpolate(Moose::FV::InterpMethod m,
-    mooseAssert(_face_info->boundaryIDs().size() == 1,
-                "I think some of my logic might depend on my implicit assumption that we have "
-                "one boundary ID at most per face");
-    mooseAssert(_flow_boundaries.find(*_face_info->boundaryIDs().begin()) != _flow_boundaries.end(),
-                "INSFV*Advection flux kernel objects should only execute on flow boundaries.");
```
I don't want this assert to go away. I really want to make sure this kernel only ever executes on a flow boundary if it's executing on a boundary. So you could do this:
```c++
#ifndef NDEBUG
bool flow_boundary_found = false;
for (const auto bid : _face_info->boundaryIDs())
  if (_flow_boundaries.find(bid) != _flow_boundaries.end())
    {
      flow_boundary_found = true;
      break;
    }

mooseAssert(flow_boundary_found, "INSFV*Advection flux kernel objects should only execute on flow boundaries.");
#endif
```

GiudGiud replied:
Sounds good. I removed it because I didnt feel like looping through all faces on every Qp. But I guess there's no other way to test where a kernel is executed

lindsayad replied:
In most cases there will only be one boundary ID to loop over so the complexity most times will be O(log(flow_boundaries.size())). And there should only ever be one qp corresponding to the face centroid. If we have more than one qp we are in trouble. 

With the ifndef guards this won’t be executed in optimized code so most users won’t notice any slow down from this sanity check. I don’t care if dbg or devel runs slowly

lindsayad commented on modules/navier_stokes/test/tests/finite_volume/ins/block_restriction/tests:
```diff
@@ -0,0 +1,10 @@
+    type = 'Exodiff'
+    input = 2d-rc.i
+    exodiff = 2d-rc_out.e
+    requirement = 'The system shall be able to restrict block-restrict all variables in a heated channel simulation with passive scalar advection.'
```
Weird I can't make a suggestion...maybe because I'm looking at this commit by commit. Anyway probably want to remove one of those "restricts".

Thank you very much for adding this test!

lindsayad commented on modules/navier_stokes/src/fvkernels/INSFVMomentumAdvection.C:
```diff
@@ -436,15 +436,8 @@ INSFVMomentumAdvection::interpolate(Moose::FV::InterpMethod m,
-              "The determineElemOneAndTwo utility determined the wrong value for elem_is_elem_one");
-  mooseAssert(elem_is_elem_one, "FVFluxKernel elem should coincide with the faceInfo elem.");
+  const Elem * const elem = &_face_info->elem();
+  const Elem * const neighbor = _face_info->neighborPtr();
```
This is actually really good. Now when people do `git blame` you will own all these lines and they will think you wrote bugs instead of me :laughing:

GiudGiud replied:
jaja. Well I ll point them to the reviewer who merged it

lindsayad reviewed (changes requested):
Just a couple minor things.

GiudGiud committed:
Address @lindsayad review

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/682012/) on 94b3f47 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16936/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16936/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format f7f66d77cd8ffeb56b76b57249cbef6402a4dbb7`

lindsayad reviewed (approved):
Failure is unrelated
